File: app/src/models/stash-entry.ts
import { CommittedFileChange } from './status'

export interface IStashEntry {
  /** The fully qualified name of the entry i.e., `refs/stash@{0}` */
  readonly name: string

  /** The name of the branch at the time the entry was created. */
  readonly branchName: string

  /** The SHA of the commit object created as a result of stashing. */
  readonly stashSha: string

  /** The list of files this stash touches */
  readonly files: StashedFileChanges

  readonly tree: string
  readonly parents: ReadonlyArray<string>
}

/** Whether file changes for a stash entry are loaded or not */
export enum StashedChangesLoadStates {
  NotLoaded = 'NotLoaded',
  Loading = 'Loading',
  Loaded = 'Loaded',
}

/**
 * The status of stashed file changes
 *
 * When the status us `Loaded` all the files associated
 * with the stash are made available.
 */
export type StashedFileChanges =
  | {
      readonly kind:
        | StashedChangesLoadStates.NotLoaded
        | StashedChangesLoadStates.Loading
    }
  | {
      readonly kind: StashedChangesLoadStates.Loaded
      readonly files: ReadonlyArray<CommittedFileChange>
    }

export type StashCallback = (stashEntry: IStashEntry) => Promise<void>

File: GitDesktop/GitDesktop/Models/StashEntry.swift
import Foundation

/// A stash entry. Port of `IStashEntry` in `models/stash-entry.ts`.
/// Files are `.notLoaded`/`.loading` until Task 8 resolves the diff.
public struct StashEntry: Sendable, Equatable, Identifiable {
    /// Fully qualified name, e.g. `refs/stash@{0}`.
    public var name: String
    public var branchName: String
    public var stashSha: String
    public var files: StashedFileChanges
    public var tree: String
    public var parents: [String]

    public init(
        name: String,
        branchName: String,
        stashSha: String,
        files: StashedFileChanges = .notLoaded,
        tree: String,
        parents: [String]
    ) {
        self.name = name
        self.branchName = branchName
        self.stashSha = stashSha
        self.files = files
        self.tree = tree
        self.parents = parents
    }

    public var id: String { name }
}

public enum StashedChangesLoadState: String, Sendable {
    case notLoaded = "NotLoaded"
    case loading = "Loading"
    case loaded = "Loaded"
}

public enum StashedFileChanges: Sendable, Equatable {
    case notLoaded
    case loading
    case loaded(files: [CommittedFileChange])

    public var kind: StashedChangesLoadState {
        switch self {
        case .notLoaded: return .notLoaded
        case .loading: return .loading
        case .loaded: return .loaded
        }
    }
}
